Lighthouse - minor improvements (#1756)
neo4j/docs-operations: modules/ROOT/pages/clustering/setup/discovery.adoc

[IMPORTANT]
====
From Neo4j 5.23, a new version v2 has been introduced.
The new version v2 is recommended for new deployments.

> 
> From Neo4j 5.23, the current version v1 has been deprecated.

Reviewer: ```suggestion
The current version v1 has been deprecated.
```

+
As an example, for those using the list resolver, the settings for all the servers should include:
+
```
dbms.cluster.discovery.resolver_type=LIST

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000
```
. Restart server01 with the new setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/in-place-1-v1-over-v2.png[]
. Run `SHOW SERVERS` and ensure that all three members are alive.
. Then repeat steps 2 and 3 for server02 and server03.
Ensure that they are set to `dbms.cluster.discovery.version=V1_OVER_V2`.
Restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v1-over-v2.png[]

. Using `bolt://`, connect to the system database of servers 1, 2, 3, and run the following procedure.
This can be done using via `./cypher-shell -a bolt://localhost:7681 -d system` for example.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> They should display "Matching" in the `stateComparison`.

Reviewer 2: ```suggestion
They should display "Matching" in the `stateComparison` column.
```

----
+
The expected result should display `v2ServerCount` as 3.
`stateComparison` is not expected to match at this stage because the original servers are not visible to the V2 discovery service.
+
.Expected result
[queryresult]
----
 +---------------------------------------------------------------------------------------------------------+
 | mode         | stateComparison                                          | v1ServerCount | v2ServerCount |
 +---------------------------------------------------------------------------------------------------------+
 | "V1_OVER_V2" | "States are not matching after PT55M36.693S: (score:29)" | "6"           | "3"           |
 +---------------------------------------------------------------------------------------------------------+
----
+
. Deallocate, drop, and shut down servers 1, 2, 3. 
+
. Start up servers 7, 8, 9, this time with the setting `dbms.cluster.discovery.version=V2_OVER_V1`.
+
image:discovery-service/v2_over_v1.png[]
+
The discovery addresses in the settings should include addresses of the new members, and the previous members.
+
As an example, for those using the list resolver, the settings for the new servers should include:
+

> ```

Reviewer 2: ```suggestion
[source, parameters]
----
```

dbms.cluster.discovery.resolver_type=LIST
dbms.cluster.discovery.version=V2_OVER_V1

dbms.cluster.discovery.endpoints=server04.example.com:5000,server05.example.com:5000,server06.example.com:5000,server07.example.com:5000,server08.example.com:5000,server09.example.com:5000
dbms.cluster.discovery.v2.endpoints=server04.example.com:6000,server05.example.com:6000,server06.example.com:6000,server07.example.com:6000,server08.example.com:6000,server09.example.com:6000

> ```

Reviewer 2: ```suggestion
----
```

+
. Using `bolt://`, connect to the system database of servers 7, 8, 9 and run the following procedure:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> The output should display `Matching` in the `stateComparison`.

Reviewer 2: ```suggestion
The output should display `Matching` in the `stateComparison` column.
```

If they are not, wait and try again till matching.
+
.Expected result
[queryresult]
----
+----------------------------------------------------------------+
| mode         | stateComparison | v1ServerCount | v2ServerCount |
+----------------------------------------------------------------+
| "V2_OVER_V1" | "Matching"      | "6"           | "6"           |
+----------------------------------------------------------------+
----
+
. Deallocate, drop, and shut down servers 4, 5, and 6.
+
. Start up servers 10, 11, 12, this time with the setting `dbms.cluster.discovery.version=V2_ONLY`.
+
image:discovery-service/v2_only.png[]
+
The discovery addresses in the settings should include addresses of the new members, and the previous members.
Note that only the v2 settings are required.
+
As an example, for those using the list resolver, the settings for the new servers should include:
+

> ```

Reviewer 2: ```suggestion
[source, parameters]
----
```

dbms.cluster.discovery.resolver_type=LIST
dbms.cluster.discovery.version=V2_ONLY

dbms.cluster.discovery.v2.endpoints=server07.example.com:6000,server08.example.com:6000,server09.example.com:6000,server10.example.com:6000,server11.example.com:6000,server12.example.com:6000

> ```

Reviewer 2: ```suggestion
----
```

+
. Deallocate, drop, and shut down servers 7, 8, 9.
+
. Finally, using `bolt://`, connect to the system database of servers 10, 11, 12, and run the following procedure to check the version of the discovery service:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
.Expected result
[queryresult]
----
+-------------------------------------------------------------+
| mode      | stateComparison | v1ServerCount | v2ServerCount |
+-------------------------------------------------------------+
| "V2_ONLY" | "N/A"           | "N/A"         | "3"           |
+-------------------------------------------------------------+
----
[[discovery-v1-to-v2-procedures]]

> 

Reviewer 2: ```suggestion
```

=== Using procedures

// By using just procedures, the nice thing is that during the user's normal version upgrade, they can also add the new settings required for v2 discovery service.
// Then, when they are ready to migrate to v2, they can use the procedures and then finally, set the version to 'V2' in the neo4j.conf at the very end - which can just sit there till the next restart
// so no lighthouse-specific server restarts are required if all is done correctly.
//Note that the settings detailed above must first be set and the servers restarted to allow the settings to take effect.

. For all the servers, ensure that new settings are updated to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.
+
As an example, for those using the list resolver, the settings for all the servers should include:
+

> ```

Reviewer 2: ```suggestion
[source, parameters]
----
```

dbms.cluster.discovery.resolver_type=LIST

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000

> ```

Reviewer 2: ```suggestion
----
```

+

> . In Cypher Shell, connect to the system database of server01 using `bolt://`.

Reviewer 2: ```suggestion
. In Cypher Shell, connect to the `system` database of server01 using `bolt://`.
```

In-place rolling reduces fault tolerance temporarily because you are restarting a running server.
To keep fault-tolerance, you can introduce a fourth server temporarily.
====

> . For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[preparation] section.

Reviewer 2: ```suggestion
. For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.
```

+
As an example, for those using the list resolver, the settings for all the servers should include:
+
```
dbms.cluster.discovery.resolver_type=LIST

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000
```
. Restart server01 with the new setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/in-place-1-v1-over-v2.png[]
. Run `Show servers;` and ensure that all three members are alive.

> . Then repeat steps 2 and 3 for server02 and server03. Ensure that they are set to `dbms.cluster.discovery.version=V1_OVER_V2`

Reviewer 2: ```suggestion
. Then repeat steps 2 and 3 for server02 and server03.
Ensure that they are set to `dbms.cluster.discovery.version=V1_OVER_V2`.
```

Restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v1-over-v2.png[]

. Using `bolt://`, connect to the system database of servers 1, 2, 3, and run the following procedure.
This can be done using via `./cypher-shell -a bolt://localhost:7681 -d system` for example.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
They should display "Matching" in the `stateComparison`.
+
[queryresult]
----
+----------------------------------------------------------------+
| mode         | stateComparison | v1ServerCount | v2ServerCount |
+----------------------------------------------------------------+
| "V1_OVER_V2" | "Matching"      | "3"           | "3"           |
+----------------------------------------------------------------+
----
+
If they are not, wait and try again.
+
. Restart server01 again with the new setting `dbms.cluster.discovery.version=V2_OVER_V1`.
+
image:discovery-service/in-place-1-v2-over-v1.png[]
+

> . Run `Show servers;` and ensure that all three members are alive.

Reviewer 2: ```suggestion
. Run `SHOW SERVERS` and ensure that all three members are alive.
```

. Then repeat steps 6 and 7 for servers 2 and 3.

> Ensure that they are set to `dbms.cluster.discovery.version=V2_OVER_V1`

Reviewer 2: ```suggestion
Ensure that they are set to `dbms.cluster.discovery.version=V2_OVER_V1`.
```

Restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v2-over-v1.png[]

. Similar to step 5, verify that `stateComparison` shows `Matching`.

. Repeat steps 6, 7, 8, 9, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`
image:discovery-service/in-place-123-v2-only.png[]
. Verify that `CALL dbms.cluster.showParallelDiscoveryState()` now shows `V2_ONLY` running.
Note that `stateComparison` is `N/A` because you do not have v1 to compare states anymore.

[[discovery-v1-to-v2-new-server]]
=== New server rolling

> . New server rolling will require 3 currently running servers, and 3 new servers.

Reviewer 2: ```suggestion
The new server rolling requires three running servers and three new servers.
```

> For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[preparation] section.

Reviewer 2: ```suggestion
. For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.
```

In-place rolling reduces fault tolerance temporarily because you are restarting a running server.
To keep fault-tolerance, you can introduce a fourth server temporarily.
====
. For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.

> +

Reviewer: Rogue `+` signs.

Reviewer 2: Yeah, there are quite a few of those. We don't need a `+` sign before a step, aka `. `.

+
. Restart server01 again with the new setting `dbms.cluster.discovery.version=V2_OVER_V1`.
+
image:discovery-service/in-place-1-v2-over-v1.png[]
+
. Run `SHOW SERVERS` and ensure that all three members are alive.
. Then repeat steps 6 and 7 for servers 2 and 3.
Ensure that they are set to `dbms.cluster.discovery.version=V2_OVER_V1`.
Restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v2-over-v1.png[]

. Similar to step 5, verify that `stateComparison` shows `Matching`.

. Repeat steps 6, 7, 8, 9, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`
image:discovery-service/in-place-123-v2-only.png[]
. Verify that `CALL dbms.cluster.showParallelDiscoveryState()` now shows `V2_ONLY` running.
Note that `stateComparison` is `N/A` because you do not have v1 to compare states anymore.

[[discovery-v1-to-v2-new-server]]
=== New server rolling
The new server rolling requires three running servers and three new servers.
. For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.

> +

Reviewer: Rogue `+` signs.

As an example, for those using the list resolver, the settings for all the servers should include:
+
```
dbms.cluster.discovery.resolver_type=LIST

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000,server04.example.com:5000,server05.example.com:5000,server06.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000,server04.example.com:6000,server05.example.com:6000,server06.example.com:6000
```
. Start up the three new servers with the setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/v1_over_v2.png[]
+
. Using `bolt://`, connect to the system database of servers 4, 5, 6, and run the following procedure.
This can be done using via `./cypher-shell -a bolt://localhost:7685 -d system` for example.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
The expected result should display `v2ServerCount` as 3.

> `stateComparison` that does show matching is fine at this point.

Reviewer: Sentence doesn't quite make sense. A suggestion:
```suggestion
`stateComparison` is not expected to match at this stage because the original servers are not visible to the V2 discovery service.
```

+
.Expected result
[queryresult]
----
 +---------------------------------------------------------------------------------------------------------+
 | mode         | stateComparison                                          | v1ServerCount | v2ServerCount |
 +---------------------------------------------------------------------------------------------------------+
 | "V1_OVER_V2" | "States are not matching after PT55M36.693S: (score:29)" | "6"           | "3"           |
 +---------------------------------------------------------------------------------------------------------+
----
+
. Deallocate, drop, and shut down servers 1, 2, 3. 

> . Run `neo4j-admin unbind` on these three servers.

Reviewer: It's not 'new server rolling' if we are re-using servers. This should be sympathetic to automatic processes which just always add a new server. So I think it's OK that it's a detailed process, that maybe has 'servers1-6' in the discovery list sometimes, sometimes 'servers4-9', etc.

Author: This makes alot of sense, it has been tweaked in [053952e](https://github.com/neo4j/docs-operations/pull/1756/commits/053952ee38ee5e55b30cea8e1a369ea3f4412f62)

Reviewer: Very nice.

In-place rolling reduces fault tolerance temporarily because you are restarting a running server.
To keep fault-tolerance, you can introduce a fourth server temporarily.
====
. For all the servers, ensure that new settings are added to the configuration as detailed the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.
+
As an example, for those using the list resolver, the settings for all the servers should include:
+

> ```

Reviewer 2: ```suggestion
[source, parameters]
----
```

Restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v2-over-v1.png[]

. Similar to step 5, verify that `stateComparison` shows `Matching`.

. Repeat steps 6, 7, 8, 9, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`
image:discovery-service/in-place-123-v2-only.png[]
. Verify that `CALL dbms.cluster.showParallelDiscoveryState()` now shows `V2_ONLY` running.
Note that `stateComparison` is `N/A` because you do not have v1 to compare states anymore.

[[discovery-v1-to-v2-new-server]]
=== New server rolling
The new server rolling requires three running servers and three new servers.

. Start up the three new servers with the setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/v1_over_v2.png[]
+
The new servers should have settings which are updated as detailed in the xref:clustering/setup/discovery.adoc#_preparation[Preparation] section.
The discovery addresses should include addresses of the new members, and the previous members.
+
As an example, for those using the list resolver, the settings for the new servers should include:
+

> ```

Reviewer 2: ```suggestion
[source, parameters]
----
```

dbms.cluster.discovery.resolver_type=LIST
dbms.cluster.discovery.version=V1_OVER_V2

dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000,server04.example.com:5000,server05.example.com:5000,server06.example.com:5000
dbms.cluster.discovery.v2.endpoints=server01.example.com:6000,server02.example.com:6000,server03.example.com:6000,server04.example.com:6000,server05.example.com:6000,server06.example.com:6000

> ```

Reviewer 2: ```suggestion
----
```

. Check the state on all servers again.
It should show that the states are `Matching`.

> Both serverCounts should be 3.

Reviewer 2: ```suggestion
Both `serverCounts` should be 3.
```
